sup3r/bias/qdm.py

```diff
@@ -495,3 +505,314 @@ def run(self,
+
+class PresRat(ZeroRateMixin, QuantileDeltaMappingCorrection):
+    """PresRat bias correction method (precipitation)
+
+    The PresRat correction [Pierce2015]_ is defined as the combination of
+    three steps:
+    * Use the model-predicted change ratio (with the CDFs);
+    * The treatment of zero-precipitation days (with the fraction of dry days);
+    * The final correction factor (K) to preserve the mean (ratio between both
+      estimated means);
+
+    References
+    ----------
+    .. [Pierce2015] Pierce, D. W., Cayan, D. R., Maurer, E. P., Abatzoglou, J.
+       T., & Hegewisch, K. C. (2015). Improved bias correction techniques for
+       hydrological simulations of climate change. Journal of Hydrometeorology,
+       16(6), 2421-2442.
+    # Todo:
+    #   - Identify Z_gf. (0.01 mm) Also have to save on output params
+    #   - Estimate K = <x> / <hat{x}>
+    """
+
+    def _init_out(self):
+        super()._init_out()
+
+        shape = (*self.bias_gid_raster.shape, 1)
+        self.out[f'{self.base_dset}_zero_rate'] = np.full(shape,
+                                                          np.nan,
+                                                          np.float32)
+        shape = (*self.bias_gid_raster.shape, 12)
+        self.out[f'{self.bias_feature}_mean_mh'] = np.full(shape,
+                                                           np.nan,
+                                                           np.float32)
+        self.out[f'{self.bias_feature}_mean_mf'] = np.full(shape,
+                                                           np.nan,
+                                                           np.float32)
+
+    # pylint: disable=W0613
+    @classmethod
+    def _run_single(
+        cls,
+        bias_data,
+        bias_fut_data,
+        base_fps,
+        bias_feature,
+        base_dset,
+        base_gid,
+        base_handler,
+        daily_reduction,
+        decimals,
+        sampling,
+        n_samples,
+        log_base,
+        *,
+        bias_ti,
+        bias_fut_ti,
+        zero_rate_threshold,
+        base_dh_inst=None,
+
+    ):
+        """Estimate probability distributions at a single site
+
+        ATTENTION: This should be refactored. There is too much redundancy in
+        the code. Let's make it work first, and optimize later.
+        """
+        base_data, base_ti = cls.get_base_data(
+            base_fps,
+            base_dset,
+            base_gid,
+            base_handler,
+            daily_reduction=daily_reduction,
+            decimals=decimals,
+            base_dh_inst=base_dh_inst,
+        )
+
+        out = cls.get_qdm_params(
+            bias_data,
+            bias_fut_data,
+            base_data,
+            bias_feature,
+            base_dset,
+            sampling,
+            n_samples,
+            log_base,
+        )
+
+        out[f'{base_dset}_zero_rate'] = cls.zero_precipitation_rate(
+            base_data,
+            zero_rate_threshold,
```

Comment:
We should add an option to calculate the zero rate threshold from the smallest non-zero observed historical value.

Reply:
This is very different to what we do now. I'll also close this one to avoid confusion, but feel free to reopen if you would like to discuss more about this.